Updated dockershim faq (#31765)
kubernetes/website: content/en/blog/_posts/2022-02-16-updated-dockershim-faq.md

---
layout: blog
title: "Updated: Dockershim Removal FAQ"
date: 2022-02-22
slug: updated-dockershim-faq
aliases: [ '/dockershim' ]
---

> This pages in an update to the [Dockershim Deprecation FAQ](/blog/2020/12/02/dockershim-faq/) created in late 2020.

Reviewer: ```suggestion
**This is an update to the original [Dockershim Deprecation FAQ](/blog/2020/12/02/dockershim-faq/) article,
published in late 2020.**
```

---
layout: blog
title: "Updated: Dockershim Removal FAQ"
date: 2022-02-22
slug: dockershim-faq
aliases: [ '/dockershim' ]
---

**This is an update to the original [Dockershim Deprecation FAQ](/blog/2020/12/02/dockershim-faq/) article,
published in late 2020.**

> This document goes over some frequently asked questions regarding the Dockershim
> deprecation [announced as a part of the Kubernetes v1.20 release](/blog/2020/12/08/kubernetes-1-20-release-announcement/). For more detail
> on the removal of Docker as a container runtime for Kubernetes kubelets, and
> what that means, check out the blog post
> [Don't Panic: Kubernetes and Docker](/blog/2020/12/02/dont-panic-kubernetes-and-docker/).

Reviewer: ```suggestion
This document goes over some frequently asked questions regarding the
deprecation and removal of _dockershim_, that was
[announced](/blog/2020/12/08/kubernetes-1-20-release-announcement/)
as a part of the Kubernetes v1.20 release. For more detail
on what that means, check out the blog post
[Don't Panic: Kubernetes and Docker](/blog/2020/12/02/dont-panic-kubernetes-and-docker/).
```
NB: _Don't Panic: Kubernetes and Docker_ is also now over a year old.

> Also, you can read [check whether Dockershim removal affects you](/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/) to determine how impactful dockershim's removal to you or your organization.

Reviewer: ```suggestion
Also, you can read [check whether dockershim removal affects you](/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)
to determine how much impact the removal of dockershim would have for you
or for your organization.
```

As the Kubernetes 1.24 release has become imminent, we've been working hard to try to make this a smooth transition.

- We've written a blog post detailing our [commitment and next steps](/blog/2022/01/07/kubernetes-is-moving-on-from-dockershim/).
- We believe there are no major blockers to migration to [other container runtimes](/docs/setup/production-environment/container-runtimes/#container-runtimes).
- There is also a [Migrating from dockershim](/docs/tasks/administer-cluster/migrating-from-dockershim/) guide available.
- We've also created a page to cover [Articles on dockershim Removal and on Using CRI-compatible Runtimes](/docs/reference/node/topics-on-dockershim-and-cri-compatible-runtimes/) that include some of the already mentioned docs as well as those from external sources (including vendor guides).

### Why is dockershim being deprecated?

> Maintaining dockershim has become a heavy burden on the Kubernetes maintainers.
> The CRI standard was created to reduce this burden and allow smooth interoperability
> of different container runtimes. Docker itself doesn't currently implement CRI,
> thus the problem.
> 
> Dockershim was always intended to be a temporary solution (hence the name: shim).
> You can read more about the community discussion and planning in the
> [Dockershim Removal Kubernetes Enhancement Proposal][drkep].

Reviewer: ```suggestion
Early versions of Kubernetes only worked with a specific container runtime:
Docker Engine. Later, Kubernetes added support for working with other container runtimes.
The CRI standard was [created](/blog/2016/12/container-runtime-interface-cri-in-kubernetes/) to
enable interoperability between orchestrators (like Kubernetes) and many different container
runtimes.
Docker Engine doesn't implement that interface (CRI), so the Kubernetes project created
special code to help with the transition, and made that _dockershim_ code part of Kubernetes
itself.

The dockershim code was always intended to be a temporary solution (hence the name: shim).
You can read more about the community discussion and planning in the
[Dockershim Removal Kubernetes Enhancement Proposal][drkep].
In fact, maintaining dockershim had become a heavy burden on the Kubernetes maintainers.
```

Additionally, features that were largely incompatible with the dockershim, such
as cgroups v2 and user namespaces are being implemented in these newer CRI
runtimes. Removing support for the dockershim will allow further development in
those areas.

[drkep]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2221-remove-dockershim

### Can I still use Docker in Kubernetes 1.20?

Yes, the only thing changing in 1.20 is a single warning log printed at [kubelet]
startup if using Docker as the runtime.

> [kubelet]: /docs/reference/command-line-tools-reference/kubelet/

Reviewer: I think it's OK to remove this question and its answer. We could instead update it to be relevant for v1.23?

If we keep it, I would revise “Docker” to “Docker Engine”.

### When will dockershim be removed?

Given the impact of this change, we are using an extended deprecation timeline.
Removal of dockershim is scheduled for Kubernetes v1.24, see [Dockershim Removal Kubernetes Enhancement Proposal][drkep].

> We will be working closely with vendors and other ecosystem groups to ensure a smooth transition and will evaluate things as the situation evolves.

Reviewer: ```suggestion
The Kubernetes project will be working closely with vendors and other ecosystem groups to ensure
a smooth transition and will evaluate things as the situation evolves.
```

> ### Can I still use dockershim after it is removed from Kubernetes?

Reviewer: ```suggestion
### Can I still use Docker Engine as my container runtime?

First off, if you use Docker on your own PC to develop or test containers: nothing changes.
You can still use Docker locally no matter what container runtime(s) you use for your
Kubernetes clusters. Containers make this kind of interoperability possible.
```

Mirantis and Docker have [committed][mirantis] to maintaining the dockershim after
it is removed from Kubernetes.

[mirantis]: https://www.mirantis.com/blog/mirantis-to-take-over-support-of-kubernetes-dockershim-2/

> ### Will my existing Docker images still work?

Reviewer: ```suggestion
### Will my existing container images still work?
```

Yes, the images produced from `docker build` will work with all CRI implementations.
All your existing images will still work exactly the same.

> ### What about private images?

Reviewer: ```suggestion
#### What about private images?
```

Yes. All CRI runtimes support the same pull secrets configuration used in

> Kubernetes, either via the PodSpec or ServiceAccount.

Reviewer: ```suggestion
Kubernetes, either via the `.spec` of a Pod, or via a ServiceAccount.
```

Reviewer: How about `Pod spec` then? I think there are more Kubernetes users who aren't familiar with the Golang-style naming than there are users who are familiar with it.

provide an end-to-end standard for managing containers.

[Open Container Initiative]: https://opencontainers.org/about/overview/
[runc]: https://github.com/opencontainers/runc
[containerd]: https://containerd.io/

### Which CRI implementation should I use?

That’s a complex question and it depends on a lot of factors. If Docker is
working for you, moving to containerd should be a relatively easy swap and
will have strictly better performance and less overhead. However, we encourage you
to explore all the options from the [CNCF landscape] in case another would be an
even better fit for your environment.

[CNCF landscape]: https://landscape.cncf.io/card-mode?category=container-runtime&grouping=category

### What should I look out for when changing CRI implementations?

While the underlying containerization code is the same between Docker and most
CRIs (including containerd), there are a few differences around the edges. Some
common things to consider when migrating are:

- Logging configuration
- Runtime resource limitations
- Node provisioning scripts that call docker or use docker via it's control socket
- Kubectl plugins that require docker CLI or the control socket

> - Kubernetes tools that require direct access to Docker (e.g. kube-imagepuller)

Reviewer: ```suggestion
- Tools from the Kubernetes project that require direct access to Docker Engine
  (for example: the deprecated `kube-imagepuller` tool)
```
(I _assume_ it's deprecated, but we should check!)

- Configuration of functionality like `registry-mirrors` and insecure registries 

> - Other support scripts or daemons that expect Docker to be available and are run

Reviewer: ```suggestion
- Other support scripts or daemons that expect Docker Engine to be available and are run
  outside of Kubernetes (for example, monitoring or security agents)
```

  outside of Kubernetes (e.g. monitoring or security agents)
- GPUs or special hardware and how they integrate with your runtime and Kubernetes

If you use Kubernetes resource requests/limits or file-based log collection
DaemonSets then they will continue to work the same, but if you’ve customized

> your dockerd configuration, you’ll need to adapt that for your new container

Reviewer: ```suggestion
your `dockerd` configuration, you’ll need to adapt that for your new container
```

runtime where possible.

Another thing to look out for is anything expecting to run for system maintenance
or nested inside a container when building images will no longer work. For the
former, you can use the [`crictl`][cr] tool as a drop-in replacement (see [mapping from docker cli to crictl](https://kubernetes.io/docs/tasks/debug-application-cluster/crictl/#mapping-from-docker-cli-to-crictl)) and for the
latter you can use newer container build options like [img], [buildah],
[kaniko], or [buildkit-cli-for-kubectl] that don’t require Docker.

[cr]: https://github.com/kubernetes-sigs/cri-tools
[img]: https://github.com/genuinetools/img
[buildah]: https://github.com/containers/buildah
[kaniko]: https://github.com/GoogleContainerTools/kaniko
[buildkit-cli-for-kubectl]: https://github.com/vmware-tanzu/buildkit-cli-for-kubectl

For containerd, you can start with their [documentation] to see what configuration
options are available as you migrate things over.

[documentation]: https://github.com/containerd/cri/blob/master/docs/registry.md

For instructions on how to use containerd and CRI-O with Kubernetes, see the
Kubernetes documentation on [Container Runtimes]

> [Container Runtimes]: /docs/setup/production-environment/container-runtimes

Reviewer: nit
```suggestion
[Container Runtimes]: /docs/setup/production-environment/container-runtimes/
```

### What if I have more questions?

If you use a vendor-supported Kubernetes distribution, you can ask them about
upgrade plans for their products. For end-user questions, please post them
to our end user community forum: https://discuss.kubernetes.io/. 

You can also check out the excellent blog post
[Wait, Docker is deprecated in Kubernetes now?][dep] a more in-depth technical
discussion of the changes.

> [dep]: https://dev.to/inductor/wait-docker-is-deprecated-in-kubernetes-now-what-do-i-do-e4m

Reviewer: How about we move that into https://kubernetes.io/docs/reference/node/topics-on-dockershim-and-cri-compatible-runtimes/ and omit it from here?

(I'm also OK with leaving it in).
